src/test_archive.cpp

```diff
@@ -0,0 +1,154 @@
+#include "engine/engine.h"
+#include "engine/archive.h"
+#include <boost/program_options.hpp>
+
+#include <string>
+#include <iostream>
+#include <cstdlib>
+#include <functional>
+#include <chrono>
+
+using namespace CityFlow;
+namespace bpo = boost::program_options;
+constexpr int interval = 100;
+
+double getDuration(const clock_t &a, const clock_t &b) {
+    return ((double)(a - b)) / CLOCKS_PER_SEC;
+}
+
+bool test0(Engine &engine);
+bool test1(Engine &engine);
+bool test2(Engine &engine);
+bool test3(Engine &engine);
+
+int main(int argc, char* argv[]) {
+
+    std::string configFile;
+    int threadNum;
+
+    bpo::options_description opts("all options");
+    bpo::variables_map vm;
+
+    opts.add_options()
+            ("configFile", bpo::value<std::string>(&configFile)->required(), "config file")
+            ("threadNum", bpo::value<int>(&threadNum)->default_value(1), "number of threads")
+            ("help", "Simulator Parameters");
+
+    try {
+        bpo::store(bpo::parse_command_line(argc, argv, opts), vm);
+        bpo::notify(vm);
+    }
+    catch (const bpo::error &e) {
+        std::cout << e.what() << std::endl;
+        return 0;
+    }
+
+    if (vm.count("help")) {
+        std::cout << opts << std::endl;
+        return 0;
+    }
+
+    std::string dataDir(std::getenv("DATADIR"));
+
+    Engine engine(dataDir + "/config/" + configFile, (unsigned int)threadNum);
+
+    std::vector<std::function<bool(Engine &engine)>> testFunctions{test0, test1, test2, test3};
+    std::vector<std::string> captions{
+        "Single save, single load",
+        "Single save, multiple loads",
+        "Multiple saves, multiple loads",
+        "Performance test"
+        };
+    for (int i = 0 ; i < interval ; ++i) engine.nextStep();
+    for (size_t testIndex = 0 ; testIndex < testFunctions.size() ; ++testIndex) {
+        std::cout <<"[test " << testIndex << "] " << captions[testIndex] << std::endl;
+        if (testFunctions[testIndex](engine))
+            std::cout << "\33[1;32m" << "PASS" << "\33[0m" << std::endl;
+        else {
+            std::cout << "\33[1;31m" << "FAIL" << "\33[0m" << std::endl;
+            return -1;
+        }
+    }
+    return 0;
+}
+
+bool test0(Engine &engine)
+{
+    auto before = engine.getLaneVehicleCount();
+    Archive archive(engine);
+    for (int i = 0 ; i < interval ; ++i) engine.nextStep();
+    auto after1 = engine.getLaneVehicleCount();
+    engine.load(archive);
+    for (int i = 0 ; i < interval ; ++i) engine.nextStep();
+    auto after2 = engine.getLaneVehicleCount();
+    return after1 == after2;
```

Comment:
I don't think it is enough that we only check `LaneVehicleCount`.

Reply:
BTW, should we add some more APIs, like `getAvarageSpeed`, `getAvarageTravelTime` or whatever? I think it is useful in both testing and RL training. @zhc134